### ISSUE TITLE
Missing '-Wreturn-type' warning for 'main' having no return statement in C89 mode

### ISSUE THREAD

**Issue by morinmorin** [llvm/llvm-project#21650]
|  |  |
| --- | --- |
| Bugzilla Link | [21276](https://llvm.org/bz21276) |
| Version | unspecified |
| OS | All |

## Extended Description 
Clang does not emit "control reaches end of non-void function [-Wreturn-type]" warning for 'main' having no return statement in C89 mode (-std=c89, gnu89). 

Clang adds implicit-return-zero in C89 mode. But, since the C89 standard does not require it, wouldn't it be better to emit warning?

**Comment by cor3ntin** [comment]
@Sirraide @AaronBallman

**Comment by AaronBallman** [comment]
Agreed, this should be given a diagnostic under `-pedantic` in C89 mode.
